src/main/scala/com/fulcrumgenomics/bam/OverlappingBasesConsensusCaller.scala

```diff
@@ -0,0 +1,250 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.bam.api.{SamRecord, SamSource}
+import com.fulcrumgenomics.commons.collection.SelfClosingIterator
+import com.fulcrumgenomics.commons.util.Logger
+import com.fulcrumgenomics.util.NumericTypes.PhredScore
+
+
+/** Consensus calls the overlapping mapped bases in read pairs.
+  *
+  * This will iterate through the mapped bases that overlap between the read and mate in a read pair.  If the read and
+  * mate agree at a given reference position, then read and mate base will not change and the base quality returned
+  * is controlled by `maxQualOnAgreement`. If they disagree at a given reference position, then the base and quality
+  * returned is controlled by `onlyMaskDisagreements`.
+  *
+  * @param onlyMaskDisagreements if the read and mate bases disagree at a given reference position, true to mask (make
+  *                              'N') the read and mate bases, otherwise pick the base with the highest base quality and
+  *                              return a base quality that's the difference between the higher and lower base quality.
+  * @param maxQualOnAgreement    if the read and mate bases agree at a given reference position, true to for the
+  *                              resulting base quality to be the maximum base quality, otherwise the sum of the base
+  *                              qualities.
+  */
+class OverlappingBasesConsensusCaller(onlyMaskDisagreements: Boolean = false,
+                                      maxQualOnAgreement: Boolean = false) {
+  private val NoCall: Byte = 'N'.toByte
+  private val NoCallQual: PhredScore = PhredScore.MinValue
+
+  private val r1BasesBuilder = Array.newBuilder[Byte]
+  private val r1QualsBuilder = Array.newBuilder[PhredScore]
+  private val r2BasesBuilder = Array.newBuilder[Byte]
+  private val r2QualsBuilder = Array.newBuilder[PhredScore]
+  private val r1Builders     = Seq(r1BasesBuilder, r1QualsBuilder)
+  private val r2Builders     = Seq(r2BasesBuilder, r2QualsBuilder)
+
+  /** Consensus calls the overlapping bases if and only if the template is a paired end where both ends map with at
+    * least one base overlapping.
+    *
+    * @param template the template to potentially correct.
+    * @return summary statistics about how many bases were examined and modified
+    */
+  def call(template: Template): CorrectionStats = (template.r1, template.r2) match {
+    case (Some(r1), Some(r2)) if r1.mapped && r2.mapped && r1.matesOverlap.contains(true) => call(r1, r2)
+    case _ => CorrectionStats(0, 0, 0, 0)
```

Comment:
Maybe make a `CorrectStats.AllZeros` constant and return that here so as to avoid creating lots of these?